docs/proposals/23-autoscaling-kube-apiserver-via-independent-hpa-and-vpa.md

```diff
@@ -0,0 +1,347 @@
+better, but still fixed, initial replica size. However, while this substantially improves efficiency, it is still less
+efficient than using VPA, as it is impossible for a single, fixed value to match the computing resource needs of a given
+cluster at all times, let alone of all clusters in the scaling class, at all times.
+
+#### 'MinChange' stabilisation prevents response to node CPU exhaustion 
+There is a known caveat in the current HVPA design, independent of what was discussed above, which has infrequent but
+severe impact on system stability:
+The vertical scaling stabilization algorithm used by HVPA revolves around a 'MinChange' concept, essentially suppressing
+vertical scaling until the recommendation differs from the current allocation (request) by said amount.
+Alone, this concept is functionally incorrect when applied to scale-up: scaling is suppressed in a state where VPA is
+indicating that the workload is starving for resources. To mitigate this problem, the existing design relies on
+resource headroom existing on the node, which can temporarily accommodate the starving workload's needs. In the field,
+this approach has been repeatedly observed to cause problems for etcd, for which a particularly high value of MinChange
+is used, as frequent evictions are highly undesirable. As an illustration, take a pod with MinChange for CPU set to
+1000 millicores, current CPU request at X millicores, actual pod need at X+2000, and available node headroom at 500.
+What happens is that actual pod usage extends as far as headroom allows, reaching X+500 and is starving for extra 1500
+millicores. Based on pod usage at X+500, VPA recommends e.g. X+900. This recommendation is below the MinChange threshold
+of X+1000. HVPA does not act, and the pod is permanently stuck starving for 1500 millicores of CPU.
+
+#### HPA-VPA policy fusion model does not account for multiplicative interaction 
+Lastly, HVPA applies an additive model to the interaction between HPA and VPA, with respect to provided computing power, but
+that interaction is actually multiplicative. As an extreme example, take a workload running at one replica with one core.
+Then consumption changes to 5 cores, leading HPA to recommend 5 single-core replicas - a change of +4 replicas. VPA
+suggests adding 4 cores to the existing single replica. A HVPA configured at 50% HPA + 50% VPA, would apply half of
+each recommendation, +2 cores and +2 replicas, resulting in (1 + 2) replicas * (1 + 2) cores. Failing to account for
+the multiplicative interaction between changes, HVPA provides a total of 9 cores to a workload which only needs 5. 
+
+### Other Benefits to Replacing the Existing Solution
+HVPA applies recommendations via direct edits to the controlled object (e.g. `Deployment`). This approach requires
+nonobvious special handling by Gardener code outside HVPA, which is error-prone.
+
```

Comment:
Why do we need this 130 lines long essay here? As a reader, I am interested in a short motivation, but then specifically (as a reviewer), how the new solution will look like (not an essay about the old solution).

Reply:
@voelzmo asked me out-of-band what my expectations are. None really (your call - I don't mind what you do with the essay here). It was a comment/feedback, that I found this passage not helpful and I'd have discussed the merits of the new approach and why we believe it will be better (on what data have we founded that proposal) rather than discussing the disadvantages of the old approach at length (beyond the initial motivation above).

Reply:
>Why do we need this 130 lines long essay here?

I don't know if their place really is here. I need them somewhere, at least per my current thinking (will expain shortly). I didn't have an external publication I could reference, and I needed to address the obvious question "HVPA declares it can be configured quite flexibly. Why are we not doing that?"

I aimed to assume as little prior HVPA knowledge from the GEP reader, as possible. For a minute, try to forget what you know about HVPA's issues, and look at HVPA's documentation and this GEP. If you skip the GEP's description of the HVPA algorighmic issues, you'll see a GEP which proposes replacing HVPA with an inferior solution. HVPA looks pretty good, per publicly available information.

Reply:
Algorithmic issues section now extracted to an independent document, and GEP references that document.